### PR TITLE
Delete linkage to enum if object gets deleted

### LINKED ISSUE #313: Removing an Object does not remove the linkage to an enum
Issue came up in https://github.com/ioBroker/ioBroker.hm-rega/issues/34

Tested by deleting an object via admin

### PR DESCRIPTION
**Pull request by foxriver76**
- via cli
- via instance removage
- via adapter methods
- fixes #313

### CODE REVIEW

**Comment by foxriver76** [comment]
Tests are failing, because I just return nothing (undefined) in callback if there is no error, but tests expect `err` to be null. So two options, adjusting tests to check for expecting err to be `not.ok` or `undefined` or should I actively return null in the cb?

**Comment by Apollon77** [comment]
I would adjust the test. In the end a falsy response should be enough, so undefined and null are ok.

**Comment by GermanBluefox** [comment]
LGTM

File: lib/adapter.js [hunk 1]
```diff
@@ -2592,13 +2592,16 @@ function Adapter(options) {
                     if (err) {
                         cb && cb(err);
                     } else if (!task.state) {
-                        setImmediate(_deleteObjects, tasks, options, cb);
+                        // if not a state, we dont have to delete state
+                        tools.removeIdFromAllEnums(adapterObjects, task.id).then(() => setImmediate(_deleteObjects, tasks, options, cb))
```
Comment (Apollon77) [inline lib/adapter.js]: Should we not continue _deleteObjects also on errors ?!

File: lib/adapter.js [hunk 2]
```diff
@@ -2592,13 +2592,16 @@ function Adapter(options) {
...
+                            .catch(e => typeof cb === 'function' && cb(e));
                     } else {
                         this.delForeignState(task.id, options, err => {
                             if (err) {
                                 cb && cb(err);
                             } else {
-                                setImmediate(_deleteObjects, tasks, options, cb);
+                                tools.removeIdFromAllEnums(adapterObjects, task.id).then(() => setImmediate(_deleteObjects, tasks, options, cb))
+                                    .catch(e => typeof cb === 'function' && cb(e));
```
Comment (Apollon77) [inline lib/adapter.js]: same here

File: lib/cli/cliObjects.js [hunk 3]
```diff
@@ -258,8 +258,11 @@ module.exports = class CLIObjects extends CLICommand {
         if (!ids || !ids.length) {
             callback && callback();
         } else {
-            objects.delObject(ids.pop, () =>
-                setImmediate(() => this._deleteObjects(objects, ids, callback)));
+            const id = ids.pop;
+            objects.delObject(id, () => {
+                tools.removeIdFromAllEnums(objects, id).then(() => setImmediate(this._deleteObjects, objects, ids, callback))
```
Comment (Apollon77) [inline lib/cli/cliObjects.js]: and here

**Comment by Apollon77** [comment]
Sorry but I have some comments ...

**Pull request merged**